### PR TITLE
Мухина Маргарита. Задача 3. Вариант 16. Сортировка Шелла с простым слиянием

### PR DESCRIPTION
**Pull request by MargaritaMuhina**
Последовательная версия 

Сортировка Шелла
Цикл for (gap = 1; gap < n / 3; gap = gap * 3 + 1); вычисляет начальный интервал (gap) для сортировки. 
Внешний цикл for (; gap > 0; gap = (gap - 1) / 3) перебирает интервалы, постепенно уменьшая их. На каждой итерации выполняется сортировка вставками, но не для всех элементов, а только для элементов, расстояние между которыми равно текущему интервалу (gap).
Внутренний цикл for (int i = gap; i < n; i++) проходит по массиву, начиная с индекса gap. Для каждого элемента sortedVec[i] выполняется следующее:
Значение элемента сохраняется во временной переменной temp.
Внутренний цикл сдвигает элементы, которые больше чем temp, на gap позиций вправо, пока не найдёт подходящее место для temp. Этот цикл является модификацией сортировки вставками: вместо сравнения с соседним элементом, сравнение идёт с элементом, находящимся на расстоянии gap.
После нахождения подходящего места, значение temp вставляется в массив.

Параллельная версия

Размер массива рассылается всем процессам с помощью broadcast.
Массив делится на части delta приблизительно равного размера, остаток распределяется между первыми процессами.
Выполняется распределение входного массива input_ между процессами.  Каждый процесс получает свою часть в local_input_.
Каждый процесс сортирует свою локальную часть с помощью shellSort.
Сбор отсортированных частей, используя merge для слияния отсортированных подмассивов, получаемых от других процессов с помощью world_.recv.

### CODE REVIEW

File: tasks/mpi/muhina_m_shell_sort/src/ops_mpi.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
+// Copyright 2023 Nesterov Alexander
+#include "mpi/muhina_m_shell_sort/include/ops_mpi.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <string>
+#include <thread>
+#include <vector>
+
+using namespace std::chrono_literals;
```
Comment (kavtorevd) [inline tasks/mpi/muhina_m_shell_sort/src/ops_mpi.cpp:10]: remove unused namespace

File: tasks/mpi/muhina_m_shell_sort/include/ops_mpi.hpp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#pragma once
+
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <boost/mpi/collectives.hpp>
+#include <boost/mpi/communicator.hpp>
+#include <memory>
+#include <numeric>
+#include <string>
```
Comment (kavtorevd) [inline tasks/mpi/muhina_m_shell_sort/include/ops_mpi.hpp:11]: remove unused include

File: tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+    muhina_m_shell_sort_mpi::ShellSortMPISequential ShellSortMPISequential(taskDataSeq);
+    ASSERT_EQ(ShellSortMPISequential.validation(), true);
+    ShellSortMPISequential.pre_processing();
+    ShellSortMPISequential.run();
+    ShellSortMPISequential.post_processing();
+
+    ASSERT_EQ(reference_min, out);
+  }
+}
```
Comment (kavtorevd) [inline tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp:182]: add test where checks empty vector

Comment (kavtorevd) [inline tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp:182]: add test where checks single-element vector

**Review by kavtorevd — CHANGES_REQUESTED** [review]

File: tasks/seq/muhina_m_shell_sort/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+std::vector<int> Get_Random_Vector(int sz, int min_value, int max_value) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<int> vec(sz);
+  for (int i = 0; i < sz; i++) {
+    vec[i] = min_value + gen() % (max_value - min_value + 1);
+  }
+  return vec;
+}
```
Comment (LizochkaSpring) [inline tasks/seq/muhina_m_shell_sort/func_tests/main.cpp:17]: Помести данную функцию в namespace, так же и в других тестах.

File: tasks/seq/muhina_m_shell_sort/src/ops_seq.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  return true;
+}
+
+bool muhina_m_shell_sort_seq::ShellSortSequential::validation() {
+  internal_order_test();
+  if (taskData->inputs_count[0] == 0) {
+    return taskData->outputs_count[0] == 0;
+  }
+  return taskData->outputs_count[0] != 0;
```
Comment (LizochkaSpring) [inline tasks/seq/muhina_m_shell_sort/src/ops_seq.cpp:44]: Пожалуйста, измени на более понятный, и добавь проверку на меньше нуля

File: tasks/seq/muhina_m_shell_sort/func_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+TEST(muhina_m_shell_sort_seq, Test_Sort_Small) {
+  std::vector<int> in = {5, 2, 9, 1, 5, 6};
+  std::vector<int> out(in.size(), 0);
+  std::vector<int> expected_result = {1, 2, 5, 5, 6, 9};
+
```
Comment (LizochkaSpring) [inline tasks/seq/muhina_m_shell_sort/func_tests/main.cpp:23]: Пожалуйста, добавь тесты на неверно веденные данные

File: tasks/mpi/muhina_m_shell_sort/src/ops_mpi.cpp (hunk 1 above)
Comment (MargaritaMuhina) [reply tasks/mpi/muhina_m_shell_sort/src/ops_mpi.cpp]: fixed

File: tasks/mpi/muhina_m_shell_sort/include/ops_mpi.hpp (hunk 2 above)
Comment (MargaritaMuhina) [reply tasks/mpi/muhina_m_shell_sort/include/ops_mpi.hpp]: fixed

File: tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp (hunk 3 above)
Comment (MargaritaMuhina) [reply tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp:242]: fixed

Comment (MargaritaMuhina) [reply tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp:242]: fixed

File: tasks/seq/muhina_m_shell_sort/func_tests/main.cpp (hunk 4 above)
Comment (MargaritaMuhina) [reply tasks/seq/muhina_m_shell_sort/func_tests/main.cpp:19]: fixed

File: tasks/seq/muhina_m_shell_sort/src/ops_seq.cpp (hunk 5 above)
Comment (MargaritaMuhina) [reply tasks/seq/muhina_m_shell_sort/src/ops_seq.cpp]: fixed

File: tasks/seq/muhina_m_shell_sort/func_tests/main.cpp (hunk 6 above)
Comment (MargaritaMuhina) [reply tasks/seq/muhina_m_shell_sort/func_tests/main.cpp:26]: fixed

**Review by LizochkaSpring — APPROVED** [review]

**Review by kavtorevd — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

File: tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+TEST(muhina_m_shell_sort, Test_Sort_NegativeValues) {
+  boost::mpi::communicator world;
+  std::vector<int> global_vec;
+  std::vector<int32_t> out;
+  std::shared_ptr<ppc::core::TaskData> taskDataPar = std::make_shared<ppc::core::TaskData>();
+
+  if (world.rank() == 0) {
+    global_vec = {-5, -2, -9, -1, -5, -6};
```
Comment (allnes) [inline tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp:171]: please generate values wirg differnet sizes

Comment (MargaritaMuhina) [reply tasks/mpi/muhina_m_shell_sort/func_tests/main.cpp:171]: fixed

**Review by allnes — APPROVED** [review]
